c10/core/MobileAllocator.cpp

```diff
@@ -0,0 +1,106 @@
+#include <c10/core/MobileAllocator.h>
+#include <c10/core/CPUAllocator.h>
+
+namespace c10 {
+namespace {
+
+struct Configuration final {
+#if defined(__arm__) || defined(__aarch64__) || defined(_M_ARM) || defined(_M_ARM64)
+  static constexpr uint32_t kAlignment = 16u;
+#else
+  static constexpr uint32_t kAlignment = gAlignment;
+#endif
+};
+
+// QNNPACK AND XNNPACK may out-of-bound access the input and / or output tensors.
+// This behavior will trigger ASAN, and may result in a segfault if the accessed
+// memory location just so happens to fall on a page the current process has no
+// read access to.  Here we define a custom allocator that allocates the extra
+// storage required to keep this behavior safe.  This allocator could have been
+// restricted to QNNPACK and XNNPACK only, but that would have negative
+// performance ramifications, as input tensors must be reallocated, and copied
+// over, if not allocated with this allocator to begin with.  Making this
+// allocator the default on mobile builds minimizes the probability of unnecessary
+// reallocations and copies.
+//
+//
+// PreGuardBytes: Number of guard bytes to allocate before the allocation.
+// PostGuardBytes: Number of guard bytes to allocate after the allocation.
+
+template <uint32_t PreGuardBytes, uint32_t PostGuardBytes>
+class DefaultMobileCPUAllocator final : public at::Allocator {
+ public:
+  DefaultMobileCPUAllocator() = default;
+  virtual ~DefaultMobileCPUAllocator() override = default;
+
+  static void deleter(void* const pointer) {
+    if (C10_UNLIKELY(!pointer)) {
+      return;
+    }
+
+    const Cast memory{pointer};
+    c10::free_cpu(memory.as_byte_ptr - PreGuardBytes);
+  }
+
+  virtual DataPtr allocate(const size_t nbytes) const override {
+    if (C10_UNLIKELY(0u == nbytes)) {
+      return {
+        nullptr,
+        nullptr,
+        &deleter,
+        at::Device(DeviceType::CPU),
+      };
+    }
+
```

Comment:
Any particular reason to special case this?  It increases code complexity and binary size.

Reply:
0 bytes would still allocate pre guard + post guard bytes when passed to alloc_cpu.  Fails a few tests that expect a nullptr returned on 0 byte allocations.

Reply:
IMO those tests are broken because malloc is allowed to return non-NULL for zero-byte allocations, but okay.